tm2/pkg/commands/ffcli/command.go

```diff
@@ -0,0 +1,278 @@
+	// invoked with its corresponding command, and its output should reflect the
+	// command's short usage, short help, and long help strings, subcommands,
+	// and available flags. Optional; if not provided, a suitable, compact
+	// default is used.
+	UsageFunc func(c *Command) string
+
+	// FlagSet associated with this command. Optional, but if none is provided,
+	// an empty FlagSet will be defined and attached during the parse phase, so
+	// that the -h flag works as expected.
+	FlagSet *flag.FlagSet
+
+	// Options provided to ff.Parse when parsing arguments for this command.
+	// Optional.
+	Options []ff.Option
+
+	// Subcommands accessible underneath (i.e. after) this command. Optional.
+	Subcommands []*Command
+
+	// A successful Parse populates these unexported fields.
+	selected *Command // the command itself (if terminal) or a subcommand
+	args     []string // args that should be passed to Run, if any
+
+	// Exec is invoked if this command has been determined to be the terminal
+	// command selected by the arguments provided to Parse or ParseAndRun. The
+	// args passed to Exec are the args left over after flags parsing. Optional.
+	//
+	// If Exec returns flag.ErrHelp, then Run (or ParseAndRun) will behave as if
+	// -h were passed and emit the complete usage output.
+	//
+	// If Exec is nil, and this command is identified as the terminal command,
+	// then Parse, Run, and ParseAndRun will all return NoExecError. Callers may
+	// check for this error and print e.g. help or usage text to the user, in
+	// effect treating some commands as just collections of subcommands, rather
+	// than being invocable themselves.
+	Exec func(ctx context.Context, args []string) error
+}
+
+// Parse the commandline arguments for this command and all sub-commands
+// recursively, defining flags along the way. If Parse returns without an error,
+// the terminal command has been successfully identified, and may be invoked by
+// calling Run.
+//
+// If the terminal command identified by Parse doesn't define an Exec function,
+// then Parse will return NoExecError.
+func (c *Command) Parse(args []string) error {
+	if c.selected != nil {
+		return nil
+	}
+
+	if c.FlagSet == nil {
+		c.FlagSet = flag.NewFlagSet(c.Name, flag.ExitOnError)
+	}
+
+	if c.UsageFunc == nil {
+		c.UsageFunc = DefaultUsageFunc
+	}
+
+	c.FlagSet.Usage = func() {
+		fmt.Fprintln(c.FlagSet.Output(), c.UsageFunc(c))
+	}
+
+	if err := ff.Parse(c.FlagSet, args, c.Options...); err != nil {
+		return err
+	}
+
+	c.args = c.FlagSet.Args()
+	if len(c.args) > 0 {
+		// Determine if ff.Parse has been interrupted by a --
+		for i := 0; i < len(args)-len(c.args); i++ {
+			if args[i] == "--" {
+				goto STOP_PARSE
+			}
+		}
+		for _, subcommand := range c.Subcommands {
+			if strings.EqualFold(c.args[0], subcommand.Name) {
+				c.selected = subcommand
+				return subcommand.Parse(c.args[1:])
+			}
+		}
+		// args[0] is not a sub command, so it's an command argument.
+		c.args = c.args[:1]
```

Comment:
Logic is unclear to me - does this not remove, in fact, the first argument as passed to Run?

Reply:
After one week of holiday, I can tell you that I have no idea :clown_face: 

That probably means the code should be improved, I'm working on a version where `ff.Parse` is invoked a single in a loop.

Reply:
See the last version, I refactored the code and put a lot of comments.